### PR TITLE
plugin pr

### PR DESCRIPTION
**Pull request by gupta-ak**

### CODE REVIEW

File: include/openenclave/bits/report.h [hunk 1]
```diff
@@ -136,6 +136,73 @@ typedef struct _oe_report
...
+**==============================================================================
+**
+** oe_evidence_header_t
+**
+**==============================================================================
+*/
+typedef struct _oe_evidence_header
+{
+    uint32_t version;
+    oe_tee_evidence_type_t tee_evidence_type; // TEE type
```
Comment (soccerGB) [inline include/openenclave/bits/report.h]: tee_evidence_type seems a bit wordy (I might have been the one created it)
How about renaming it to evidence_type?
Same comment applies to tee_evidence_size?

File: samples/attestation_plugin/enclave/enc.c [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    size_t claim_count = 0;
+    uint8_t* user_data1 = NULL;
+    size_t user_data_size1 = 0;
+    oe_quote_customization_plugin_context_t* plugin1 = create_oe_plugin();
+
+    fprintf(stdout, "Hello from enclave::enclave_attestation_plugin\n");
+
+    // Register attestation plugins.
+    result = oe_register_attestation_plugin(plugin1, NULL, 0);
+    if (result != OE_OK)
```
Comment (soccerGB) [inline samples/attestation_plugin/enclave/enc.c:34]: It would be interesting to have sample that has no null config data

File: enclave/attest_plugin.c [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+    total_evidence_buf = (uint8_t*)oe_malloc(total_evidence_size);
+    if (total_evidence_buf == NULL)
+        OE_RAISE(OE_OUT_OF_MEMORY);
+
+    // TODO(akagup): Fix oe report header to only have 1 size parameter.
+    header = (oe_evidence_header_t*)total_evidence_buf;
+    header->version = OE_REPORT_HEADER_VERSION;
+    header->evidence_format_uuid = *evidence_format_uuid;
+    header->tee_evidence_type = OE_TEE_TYPE_CUSTOM;
+    header->tee_evidence_size = 0;
```
Comment (soccerGB) [inline enclave/attest_plugin.c]: all the plug-in will have CUSTOM as evidence type?
Is there anyway for a plugin to support SGX or Cyres type?

File: include/openenclave/internal/report.h [hunk 4]
```diff
@@ -47,38 +48,10 @@ typedef struct _oe_get_qe_identity_info_args
...
-    uint8_t report[];
-} oe_report_header_t;
-
-OE_STATIC_ASSERT(sizeof(oe_report_header_t) == 16);
-OE_STATIC_ASSERT(
-    OE_OFFSETOF(oe_report_header_t, report) == sizeof(oe_report_header_t));
+// OE_STATIC_ASSERT(sizeof(oe_evidence_header_t) == 64);
+// OE_STATIC_ASSERT(
+//     OE_OFFSETOF(oe_evidence_header_t, report) ==
+//     sizeof(oe_evidence_header_t));
```
Comment (soccerGB) [inline include/openenclave/internal/report.h]: This block is no longer needed

File: samples/attestation_plugin/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,35 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+cmake_minimum_required(VERSION 3.11)
+
+project("Hello World Sample" LANGUAGES C CXX)
+
```
Comment (soccerGB) [inline samples/attestation_plugin/CMakeLists.txt]: sample name needs to be updated

File: samples/attestation_plugin/README.md [hunk 6]
```diff
@@ -0,0 +1,528 @@
+
+# The attestation_plugin sample
+
+- Written in C
+- Minimum code needed for an Open Enclave app
+- Help understand the basic components an OE(Open Enclave) application
+- Demonstrate how to build, sign, and run an OE image
+- Also runs in OE simulation mode
+
```
Comment (soccerGB) [inline samples/attestation_plugin/README.md:9]: Above block needs to be updated

File: samples/attestation_plugin/enclave/enc.c [hunk 7]
```diff
@@ -0,0 +1,116 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// TODO:
```
Comment (soccerGB) [inline samples/attestation_plugin/enclave/enc.c]: this block is no longer relevant

File: samples/attestation_plugin/enclave/enc.c [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    }
+    fprintf(
+        stdout,
+        "oe_register_attestation_plugin succeeded for "
+        "evidence_format_uuid (1)\n");
+
+    result = oe_get_attestation_evidence(
+        &plugin1->evidence_format_uuid,
+        (const uint8_t*)"Hello World!",
+        sizeof("Hello World!") - 1,
```
Comment (soccerGB) [inline samples/attestation_plugin/enclave/enc.c]: why minus 1 instead +1?

File: samples/attestation_plugin/enclave/plugin1/plugin1.c [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+
+    ret = 0;
+
+done:
+    mbedtls_sha256_free(&ctx);
+    return ret;
+}
+
+static int oe_get_evidence(
+    oe_quote_customization_plugin_context_t* plugin_context,
```
Comment (soccerGB) [inline samples/attestation_plugin/enclave/plugin1/plugin1.c]: another name without oe_ would be less confusing even it's already declared as "static"

File: samples/attestation_plugin/enclave/plugin1/plugin1.c [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+static int oe_verify_evidence(
+    oe_quote_customization_plugin_context_t* plugin_context,
+    const uint8_t* evidence_buffer,
+    size_t evidence_buffer_size,
+    oe_claim_element_t** claims,
+    size_t* claim_count,
+    uint8_t** user_data,
+    size_t* user_data_size)
+{
+    return 0;
```
Comment (soccerGB) [inline samples/attestation_plugin/enclave/plugin1/plugin1.c]: It would be great to do something here to make this sample more interesting.

File: samples/attestation_plugin/enclave/plugin2/plugin2.c [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+    // TODO:
+    // verify the Sample JWT token here
+
+    ret = 0;
+done:
+    return ret;
+}
+
+static oe_quote_customization_plugin_callbacks_t attestation_callbacks = {
```
Comment (soccerGB) [inline samples/attestation_plugin/enclave/plugin2/plugin2.c]: I got a feeling that this plugin is no longer relevant. Right?

File: samples/attestation_plugin/enclave/plugin3/plugin3.c [hunk 12]
```diff
@@ -0,0 +1,188 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (soccerGB) [inline samples/attestation_plugin/enclave/plugin3/plugin3.c]: This plugin3.c seems to be irrelevant.

it would be great if this is replaced by some sample JWT-like token validation plugin.

File: include/openenclave/bits/report.h (hunk 1 above)
Comment (gupta-ak) [reply include/openenclave/bits/report.h]: Agreed, most of the names are pretty wordy. Honestly, it could just be `type` considering that the type name already tells you what it is.

File: samples/attestation_plugin/enclave/enc.c (hunk 2 above)
Comment (gupta-ak) [reply samples/attestation_plugin/enclave/enc.c:34]: Yeah, I've only partially update sample 1.

I think sample 2 could be something like a challenge / response plugin with some config data and sample 3 could be some custom format.

File: samples/attestation_plugin/enclave/plugin2/plugin2.c (hunk 11 above)
Comment (gupta-ak) [reply samples/attestation_plugin/enclave/plugin2/plugin2.c]: Yeah. I didn't delete samples 2 and 3, because I was planning to update them.

File: samples/attestation_plugin/enclave/plugin3/plugin3.c (hunk 12 above)
Comment (gupta-ak) [reply samples/attestation_plugin/enclave/plugin3/plugin3.c]: Yeah. I didn't delete samples 2 and 3, because I was planning to update them.

File: samples/attestation_plugin/enclave/plugin1/plugin1.c (hunk 10 above)
Comment (gupta-ak) [reply samples/attestation_plugin/enclave/plugin1/plugin1.c]: Yeah, I haven't finished sample 1 yet, since we had some discussions about attestation this week. I'm was planning on it do the verification like how the report generation is implemented.

File: samples/attestation_plugin/enclave/plugin1/plugin1.c (hunk 9 above)
Comment (gupta-ak) [reply samples/attestation_plugin/enclave/plugin1/plugin1.c]: Sounds good.

File: samples/attestation_plugin/enclave/enc.c (hunk 8 above)
Comment (gupta-ak) [reply samples/attestation_plugin/enclave/enc.c]: sizeof("Hello World!") will return the string length and +1 for the null, so -1 to ignore the null terminator.

File: samples/attestation_plugin/enclave/enc.c (hunk 7 above)
Comment (gupta-ak) [reply samples/attestation_plugin/enclave/enc.c]: I'll remove.

File: samples/attestation_plugin/CMakeLists.txt (hunk 5 above)
Comment (gupta-ak) [reply samples/attestation_plugin/CMakeLists.txt]: I'll update the documentation on the samples once we finalize what we want the APIs to be.

File: include/openenclave/internal/report.h (hunk 4 above)
Comment (gupta-ak) [reply include/openenclave/internal/report.h]: Thanks, I'll rtemove this.

File: enclave/attest_plugin.c (hunk 3 above)
Comment (gupta-ak) [reply enclave/attest_plugin.c]: I made them all custom because I wasn't sure what a plugin w/ a SGX/Cyres type meant. Is that just a plugin that supports SGX only?

Comment (soccerGB) [reply enclave/attest_plugin.c]: Per our earlier discussion, let's keep it as custom as it is.

File: include/openenclave/attestation_plugin.h [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+/**
+ * oe_free_claims_list
+ *
+ * Frees a claims list.
+ *
+ * @param[in] claims The claims list.
+ * @param[in] claims_count The number of claims in the claims list.
+ */
+void oe_free_claim_list(oe_claim_element_t* claims, size_t claims_count);
+
```
Comment (soccerGB) [inline include/openenclave/attestation_plugin.h:251]: Function name and comment does not match

File: include/openenclave/bits/report.h (hunk 1 above)
Comment (soccerGB) [reply include/openenclave/bits/report.h]: "type" works for me.

File: include/openenclave/internal/report.h (hunk 4 above)
Comment (soccerGB) [reply include/openenclave/internal/report.h]: Thanks

File: samples/attestation_plugin/enclave/enc.c (hunk 2 above)
Comment (soccerGB) [reply samples/attestation_plugin/enclave/enc.c:34]: That's a good idea to have that in a sample

File: samples/attestation_plugin/enclave/enc.c (hunk 8 above)
Comment (soccerGB) [reply samples/attestation_plugin/enclave/enc.c]: I thought you want to send over the null terminator. (In my generic MPC sample, when I printed out the received string without terminating null, it was quite ugly in the output :-)

If you don;t want to send over null terminator, it would be great to add some comment on it.
